source/robot_model/src/Model.cpp

```diff
@@ -29,17 +39,175 @@ bool Model::create_urdf_from_string(const std::string& urdf_string, const std::s
+    std::set<std::string> extracted_words;
+    std::regex package_path_pattern(R"(filename=\"([^\"]+)\")");
+    std::regex package_name_pattern(R"(package://([^/]+)/)");
+
+    std::smatch matches_file;
+    std::smatch matches_package_name;
+    auto start = this->urdf_.str().cbegin();
+    auto end = this->urdf_.str().cend();
+
+    // Extract package paths
+    while (std::regex_search(start, end, matches_file, package_path_pattern)) {
+        std::string path = matches_file[1];
+        
+        if (std::regex_search(path, matches_package_name, package_name_pattern)) {
+            extracted_words.insert(matches_package_name[1]);
+        }
+        
+        start = matches_file[0].second;   
+    }
+    return extracted_words;
+}
+
+void Model::replace_package_by_full_path(const std::string& package_name, const std::string& full_path) {
+    std::string target = "package://" + package_name + "/";
+    std::string replacement = full_path;
+
+    std::string urdf = this->urdf_.str();
+
+    // Start from the beginning of the result string
+    size_t start_position = 0;
+    while ((start_position = urdf.find(target, start_position)) != std::string::npos) {
+        // Replace the target with the replacement string
+        urdf.replace(start_position, target.length(), replacement);
+        // Move past the last replacement
+        start_position += replacement.length();
+    }
+    this->urdf_.str(urdf);
+}
+
+void Model::resolve_package_paths() {
+    std::set<std::string> package_names = this->extract_package_name_from_urdf();
+
+    for (const auto& package_name : package_names) {
+        std::string package_path = this->meshloader_callback_(package_name);
+        this->package_paths_.push_back(package_path);
+        this->replace_package_by_full_path(package_name, package_path);
+    }
+}
+
+// read the URDF file and return the string
+std::stringstream Model::read_urdf_from_file() {
+    std::ifstream file_stream(this->get_urdf_path());
+    if (!file_stream.is_open()) {
+        throw std::runtime_error("Unable to open file: " + this->get_urdf_path());
+    }
+    std::stringstream buffer;
+    buffer << file_stream.rdbuf(); // Read the file content into the string stream
+
+    return buffer; // Convert the string stream to a string and return
+}
+
 void Model::init_model() {
-  pinocchio::urdf::buildModel(this->get_urdf_path(), this->robot_model_);
+  // initialize the collision package paths to empty
+  this->package_paths_ = {};
+
+  // check if the package path resolver callback is set then build model with the package resolver
+  if (this->load_collision_geometries_) { 
+    // read the URDF from file
+    this->urdf_ = this->read_urdf_from_file();
+
+    // resolve the package paths
+    this->resolve_package_paths();
+
+    // build the model with the resolved URDF
+    pinocchio::urdf::buildModelFromXML(this->urdf_.str(), this->robot_model_);
+
+    this->init_geom_model();
```

Comment:
rather than having a class property `package_paths_` that you update in different places, you could just return it from `resolve_package_paths()` and then pass it to `init_geom_model(...)` as an argument.

```suggestion
    // resolve the package paths
    auto package_paths = this->resolve_package_paths();

    // build the model with the resolved URDF
    pinocchio::urdf::buildModelFromXML(this->urdf_.str(), this->robot_model_);

    this->init_geom_model(package_paths);
```

It is generally advisable to keep a variable to the lowest scope possible. Here, using a class property is a bit like using a global variable. You already saw that you needed to take extra precautions in case of unintended sequence breaks (for example when you do `// initialize the collision package paths to empty: this->package_paths_ = {};`). So it's better just to narrow the scope.